strawberryfields/tdm/tdmprogram.py

```diff
@@ -0,0 +1,232 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+This module implements the :class:`.TDMProgram` class which acts as a representation for time-domain quantum circuits.
+"""
+# pylint: disable=too-many-instance-attributes,attribute-defined-outside-init
+
+from operator import itemgetter
+from math import ceil
+import strawberryfields as sf
+from strawberryfields import ops
+from strawberryfields.parameters import par_is_symbolic
+
+
+def shift_by(l, n):
+    """Convenience function to shift a list by a number of steps.
+
+    If ``n`` is positive it shifts to the left.
+
+    Args:
+        l (list): list to be shifted
+        n (int): size of the shift
+    """
+    return l[n:] + l[:n]
+
+
+def get_modes(cmd, q):
+    """Returns the registers from q required by a command"""
+    reg_getter = itemgetter(*[r.ind for r in cmd.reg])
+    modes = reg_getter(q)
+
+    if not isinstance(modes, tuple):
+        modes = (modes,)
+
+    return modes
+
+
+def validate_measurements(circuit, N):
+    """Validate the TDM program measurements are correct"""
+    spatial_modes = 0
+    measurement_reg = []
+
+    for cmd in circuit:
+        if isinstance(cmd.op, ops.Measurement):
+            modes = [r.ind for r in cmd.reg]
+            measurement_reg.extend(modes)
+            spatial_modes += len(modes)
+
+    if not spatial_modes:
+        raise ValueError("Must be at least one measurement.")
+
+    if spatial_modes is not len(N):
+        raise ValueError("Number of measurement operators must match number of spatial modes.")
+
+    return spatial_modes
+
+
+def input_check(args, copies):
+    """Checks the input arguments have consistent dimensions"""
+    # check if all lists are of equal length
+    param_lengths = [len(param) for param in args]
+    if len(set(param_lengths)) != 1:
+        raise ValueError("Gate-parameter lists must be of equal length.")
+
+    # check copies
+    if not isinstance(copies, int) or copies < 1:
+        raise TypeError("Number of copies must be a positive integer.")
+
+
+def _get_mode_order(num_of_values, N):
+    """Get the order by which the modes where measured"""
+    N = N.copy()
+
+    all_modes = []
+    for i in range(len(N)):
+        ra = list(range(sum(N[:i]), sum(N[: i + 1])))
+        all_modes.append(ra * ceil(max(N) / len(ra)))
+
+    mode_order = [i for j in zip(*all_modes) for i in j]
+    mode_order *= ceil(num_of_values / len(mode_order))
+
+    return mode_order[:num_of_values]
+
+
+def reshape_samples(all_samples, modes, N):
+    """Reshapes the samples dict so that they have the expected correct shape
+
+    Reshapes the samples dict so that they the measured modes are the ones
+    defined to be measured in the circuit instead of being spread over a larger
+    number of modes due to the mode-shifting occuring in TDMProgram.
```

Comment:
```suggestion
    defined to be measured in the circuit, instead of being spread over a larger
    number of modes due to the mode-shifting occurring in :class:`~.TDMProgram`.
```